feat: add `User` class to manage and create users within Argilla

alvarobartt opened the pull request:
# Description

This PR addresses the addition of a new user-management layer in the Python client via the `User` class, wrapping up the `/api/users` and `/api/me` endpoints in the API v0.

Closes #2968

**Type of change**

- [X] New feature (non-breaking change which adds functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

- [x] Add unit tests for the new `User` class
- [X] Add unit tests for the `user_api` endpoints
- [X] Revisit existing tests to pass with updated `UserModel` (formerly named `User`)

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

alvarobartt committed:
Curate `docstring` typos & redundancy

alvarobartt committed:
Re-define `whoami` to expect a `HttpxClient`

alvarobartt committed:
Fix `whoami` unit tests

Improve `whoami` mock in `conftest`, and remove duplicated client-error tests

alvarobartt committed:
Remove string-strip leftovers (`" " `)

alvarobartt committed:
Add `UserModel` and `UserCreateModel`

Aligned with API, as before there was no alignment due to the lack of an integration within the Python client

alvarobartt committed:
Add `list_users`, `create_user`, and `delete_user`

alvarobartt committed:
Remove unnecessary type-hint

alvarobartt committed:
Fix tests using outdated `User` & `whoami`

alvarobartt committed:
feat(wip): add `User` & `list_users`

alvarobartt committed:
Merge branch 'develop' of github.com:argilla-io/argilla into feat/user-management

alvarobartt committed:
test: re-use `mock_init_ok` in `test_aligned_argilla_versions`

alvarobartt committed:
fix: rename `ignore_none` to `exclude_none`

alvarobartt committed:
fix: expect HTTP 200 in `create_user`

alvarobartt committed:
feat: add `User` class to manage existing users

alvarobartt committed:
feat: add `User` to `__init__`

Meant to be imported as `rg.User`

davidberenstein1957 commented on src/argilla/client/users.py:
```diff
@@ -0,0 +1,118 @@
+        return None
+
+    def __str__(self) -> str:
+        return f"User(id={self.id}, username={self.username}, role={self.role}, workspaces={self.workspaces}, api_key={self.api_key}, first_name={self.first_name}, last_name={self.last_name}, role={self.role}, inserted_at={self.inserted_at}, updated_at={self.updated_at})"
```
Perhaps this can formatted a bit better to make it more readable within newlines??

alvarobartt replied:
Sure, I'll do so whenever I come back to this PR 👍🏻

davidberenstein1957 commented on src/argilla/client/sdk/users/models.py:
```diff
@@ -12,17 +12,35 @@
+    last_name: Optional[str] = Field(min_length=1)
+    username: str = Field(min_length=1, regex=r"^(?!-|_)[a-z0-9-_]+$")
+    role: Roles = "annotator"
+    password: str = Field(min_length=8, max_length=100)
```
isn't 100 a bit much for as password?

alvarobartt replied:
That's the API default, so we should be compliant with it. Anyway feel free to open a discussion with @frascuchon and/or @gabrielmbmb 😄

gabrielmbmb replied:
I think 100 is OK! 1Password generates passwords up to 100 characters.

davidberenstein1957 commented on src/argilla/client/sdk/users/api.py:
```diff
@@ -12,10 +12,147 @@
+
+    response = client.delete(url=url)
+
+    if response.status_code == 200:
```
maybe we can simplify this a bit because it is duplicated a couple of time.

alvarobartt replied:
We will refactor this at some point if applicable yes, but closing this for now!

davidberenstein1957 commented on src/argilla/client/users.py:
```diff
@@ -0,0 +1,118 @@
+        role: Optional[Roles] = None,
+    ) -> "User":
+        if not first_name:
+            warnings.warn(
```
does `warn.warn` include rich formatting?

alvarobartt replied:
No it does not, but we decided to use both `logger.warning` and `warnings.warn`, as those are intended to different purposes, see https://docs.python.org/3/howto/logging.html#when-to-use-logging

<img width="658" alt="image" src="https://github.com/argilla-io/argilla/assets/36760800/489747ac-8a29-4544-97eb-7963d0f4ccae">

alvarobartt replied:
In this case, defining a `first_name` makes the warning disappear, so this is why here we're using `warnings.warn` instead of `logging.warning`, anyway we can have a more in-detail discussing for this matter! 👍🏻

davidberenstein1957 commented on src/argilla/client/users.py:
```diff
@@ -0,0 +1,118 @@
+            role = "annotator"
+
+        if isinstance(password, UUID):
+            warnings.warn("Since the `password` is a UUID, it will be converted to a string.")
```
Maybe this is irrelevant since the str(UUID) looks exactly the same right?

alvarobartt replied:
Remove it! Thanks 👍🏻

alvarobartt committed:
Merge branch 'develop' into feat/user-management

alvarobartt committed:
fix: avoid `BaseModel.construct` to keep validation

alvarobartt committed:
fix(test): wrong type-hint for `argilla_user`

alvarobartt committed:
fix(test): diff between `UserModel` and `User`

alvarobartt committed:
chore: remove unused imports

alvarobartt committed:
fix: use `http_client` instead of `_client` in `Argilla`

alvarobartt committed:
fix(test): mock `whoami` with valid client

alvarobartt committed:
fix(test): users SDK unit tests

alvarobartt committed:
feat(wip): align `User` with `Workspace` class

alvarobartt committed:
Merge branch 'develop' of github.com:argilla-io/argilla into feat/user-management

alvarobartt committed:
Merge branch 'develop' of github.com:argilla-io/argilla into feat/user-management

alvarobartt committed:
revert: changes in `test/conftest.py`

alvarobartt committed:
revert: use `AuthenticatedClient` for `whoami`

alvarobartt committed:
revert: `whoami` alignment with SDK functions

alvarobartt committed:
fix: import `AuthenticatedClient` from `argilla.client.sdk.client`

alvarobartt committed:
feat: add `httpx` arg in `__active_client`

To be removed once all the SDK functions are aligned to expect the `httpx.Client` to be provided instead of the `AuthenticatedClient`

alvarobartt committed:
fix: use `UserRole` enum instead of `Literal`

alvarobartt committed:
test: add `/api/users` sdk unit tests

alvarobartt committed:
docs: update `CHANGELOG.md`

alvarobartt committed:
fix: `delete_user` arg is `user_id` not `id`

alvarobartt committed:
fix: exception handling in `User`

alvarobartt committed:
test: add `User` unit tests

alvarobartt committed:
fix: remove outdated code-comment

gabrielmbmb commented on src/argilla/client/sdk/users/api.py:
```diff
@@ -12,10 +12,155 @@
+        A `UserModel` instance with the current user information.
+    # Returns:
+    #     A `Response` object containing a `parsed` attribute with the parsed response if
+    #     the request was successful, which is an instance of `UserModel`.
```
I guess that we will use this once we update the function to return `Response` instead of `UserModel`. In the meantime, can we remove or move this comment to inside of the function? It will get rendered in the user docstring.

alvarobartt replied:
True I'll remove everything in the same commit to easily revert it later!

frascuchon commented on src/argilla/client/sdk/users/models.py:
```diff
@@ -12,17 +12,35 @@
-from pydantic import BaseModel
+from pydantic import BaseModel, Field
 
+from argilla.server.models import UserRole as Roles
```
I think you should have your own UserRole class instead of using it from the server module

alvarobartt replied:
I agree! Thanks

frascuchon commented on src/argilla/client/sdk/users/models.py:
```diff
@@ -12,17 +12,35 @@
+    first_name: str = Field(min_length=1)
+    last_name: Optional[str] = Field(min_length=1)
+    username: str = Field(min_length=1, regex=r"^(?!-|_)[a-z0-9-_]+$")
+    role: Roles = "annotator"
```
```suggestion
    role: Roles = Roles.annotator
```

alvarobartt replied:
Regarding how `Enum`s are accessed in Python 3.11, do you know if this may eventually lead to any issue? cc @gabrielmbmb

gabrielmbmb commented on src/argilla/client/users.py:
```diff
@@ -0,0 +1,310 @@
+            f" inserted_at={self.inserted_at}, updated_at={self.updated_at})"
+        )
+
+    @staticmethod
+    def __active_client(httpx: bool = True) -> Union["httpx.Client", "AuthenticatedClient"]:
+        """Returns the active Argilla `httpx.Client` instance."""
+        try:
+            client = active_client().http_client
+            if httpx:
+                client = client.httpx
+            return client
+        except Exception as e:
+            raise RuntimeError(f"The `rg.active_client()` is not available or not respoding.") from e
```
I think that we should create a base class that could be used in client classes that have to use the `httpx` client.

alvarobartt replied:
Sure, we can do that on another PR, to also make sure the most recent classes are updated and aligned i.e. `FeedbackDataset`, `User` and `Workspace`

frascuchon commented on src/argilla/client/users.py:
```diff
@@ -0,0 +1,310 @@
+)
+from argilla.client.sdk.users import api as users_api
+from argilla.client.sdk.users.models import UserCreateModel, UserModel
+from argilla.server.models import UserRole
```
Same here, this class definition should be in the `sdk.users.models` module

alvarobartt replied:
I agree, it was something temporary that I forgot to update!

gabrielmbmb commented on src/argilla/client/users.py:
```diff
@@ -0,0 +1,310 @@
+            raise RuntimeError(f"Error while deleting user with username=`{self.username}` from Argilla.") from e
+
+    @classmethod
+    def __new_instance(cls, client: Optional["httpx.Client"] = None, ws: Optional["UserModel"] = None) -> "User":
```
does `ws` stand for workspace user?

alvarobartt replied:
Nope that was a copy-paste issue from the `__new_instance` method defined in `Workspace` 😄

frascuchon commented on src/argilla/client/users.py:
```diff
@@ -0,0 +1,310 @@
+        """Returns a new `User` instance."""
+        instance = cls.__new__(cls)
+        instance.__client = client or cls.__active_client()
+        if isinstance(ws, UserModel):
```
```suggestion
        if isinstance(user, UserModel):
```

frascuchon commented on src/argilla/client/users.py:
```diff
@@ -0,0 +1,310 @@
+            raise RuntimeError(f"Error while deleting user with username=`{self.username}` from Argilla.") from e
+
+    @classmethod
+    def __new_instance(cls, client: Optional["httpx.Client"] = None, ws: Optional["UserModel"] = None) -> "User":
```
```suggestion
    def __new_instance(cls, client: Optional["httpx.Client"] = None, user: Optional["UserModel"] = None) -> "User":
```

frascuchon commented on tests/client/sdk/users/test_api.py:
```diff
@@ -12,25 +12,103 @@
-def test_whoami_with_connection_error():
-    with pytest.raises(BaseClientError):
-        whoami(AuthenticatedClient(base_url="http://localhost:6900", token="wrong-apikey"))
+def test_list_users(owner: User) -> None:
+    UserFactory.create(username="user_1")
+    UserFactory.create(username="user_2")
+    httpx_client = ArgillaSingleton.init(api_key=owner.api_key).http_client.httpx
+
+    response = list_users(client=httpx_client)
```
For this kind of tests I think you can use directly the `client: TestClient`, fixture
```suggestion
def test_list_users(client: TestClient, owner: User, owner_auth_header: dict) -> None:
    UserFactory.create(username="user_1")
    UserFactory.create(username="user_2")
    client.headers.update(owner_auth_header)

    response = list_users(client=client)
```

alvarobartt replied:
But won't the `ArgillaSingleton` use the `TestClient` as configured in the `conftest.py`? Is this change worth it? Asking because if so, we should change this is more than one place, LMKWYT!

frascuchon replied:
Yes, just to be more explicit here in tests if we're using the HTTP client directly. The conftest is a fixture to use the rg.init functions with an already mocked client

frascuchon commented on tests/client/test_users.py:
```diff
@@ -0,0 +1,92 @@
+        User(id="00000000-0000-0000-0000-000000000000")
+
+
+def test_user_from_name(owner: UserModel):
```
Maybe is more clear as ServerUser
```suggestion
def test_user_from_name(owner: ServerUser):
```

frascuchon commented on tests/client/test_users.py:
```diff
@@ -0,0 +1,92 @@
+
+def test_user_from_id(owner: UserModel):
+    new_user = UserFactory.create(username="test_user")
+    ArgillaSingleton.init(api_key=owner.api_key)
```
It should work with the general `rg.init` function. Just to be more familiar with inits

alvarobartt replied:
Yes, it was basically to avoid using `rg` in the unit tests and use the actual class instead, which seems more clear when debugging to have the direct import instead of the reference to `rg`

frascuchon replied:
we should align then the rest of the tests ( we can tackle in a separate PR)

alvarobartt committed:
fix: apply suggestions from code review

Co-authored-by: Gabriel Martin <gabriel@argilla.io>
Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
revert(test): mock `whoami` response

alvarobartt committed:
style: rename `UserModel`/`User` to `ServerUser`

Also included under `TYPE_CHECKING` when applicable

Co-authored-by: Francisco Aranda <francis@argilla.io>

frascuchon commented on src/argilla/client/sdk/users/models.py:
```diff
@@ -12,17 +12,40 @@
 
-class User(BaseModel):
-    """User data model"""
+class UserRoles(str, Enum):
```
Normally this enun class definition is named in singular since attributes and properties will have one single value of this class so one `UserRole`

frascuchon commented on src/argilla/client/sdk/users/models.py:
```diff
@@ -12,17 +12,40 @@
+    first_name: str = Field(min_length=1)
+    last_name: Optional[str] = Field(min_length=1)
+    username: str = Field(min_length=1, regex=r"^(?!-|_)[a-z0-9-_]+$")
+    role: UserRoles = UserRoles.annotator
```
I think is better as `role: UserRole`

alvarobartt committed:
style: rename `UserRoles` enum to `UserRole`

Co-authored-by: Francisco Aranda <francis@argilla.io>
